pkg/rules/rules.go

```diff
@@ -15,27 +16,137 @@ import (
 // Retriever allows to retrieve rules from gRPC server implementation.
 // TODO(bwplotka): Switch to native gRPC transparent client->server adapter once available.
 type Retriever struct {
-	proxy rulespb.RulesServer
+	proxy                 rulespb.RulesServer
+	replicaLabels         map[string]struct{}
+	enablePartialResponse bool
 }
 
-func NewRetriever(rs rulespb.RulesServer) *Retriever {
-	return &Retriever{
-		proxy: rs,
+func NewRetriever(rs rulespb.RulesServer, replicaLabels []string, enablePartialResponse bool) *Retriever {
+	rr := &Retriever{
+		proxy:                 rs,
+		replicaLabels:         map[string]struct{}{},
+		enablePartialResponse: enablePartialResponse,
 	}
+
+	for _, label := range replicaLabels {
+		rr.replicaLabels[label] = struct{}{}
+	}
+
+	return rr
 }
 
 func (rr *Retriever) RuleGroups(ctx context.Context) ([]*rulespb.RuleGroup, storage.Warnings, error) {
 	resp := &rulesServer{ctx: ctx}
 
+	strategy := storepb.PartialResponseStrategy_ABORT
+
+	if rr.enablePartialResponse {
+		strategy = storepb.PartialResponseStrategy_WARN
+	}
+
 	if err := rr.proxy.Rules(&rulespb.RulesRequest{
-		PartialResponseStrategy: storepb.PartialResponseStrategy_ABORT,
+		PartialResponseStrategy: strategy,
 	}, resp); err != nil {
 		return nil, nil, errors.Wrap(err, "proxy RuleGroups()")
```

Comment:
```suggestion
		return nil, nil, errors.Wrap(err, "gRPC Rules")
```

Reply:
The key part is that we want to tell what operation errored, not where are we now.